Assets/Scripts/Enemies/BaseEnemyAI.cs

```diff
@@ -0,0 +1,156 @@
+            horizontalMove = (player.transform.position - transform.position).normalized.x;
+            if (horizontalMove > 0 && !isFacingRight)
+            {
+                Flip();
+            }
+            else if (horizontalMove < 0 && isFacingRight)
+            {
+                Flip();
+            }
+            if (Time.time >= nextProjectileTime)
+            {
+                nextProjectileTime = Time.time + projectileAttackData.duration;
+                animator.SetTrigger(Constants.PROJECTILE_ATTACK_ANIMATION);
+            }
+        }
+        else
+        {
+            if (distanceToPlayer < meleeRange && Time.time >= nextMeleeTime && !isFiringProjectile)
+            {
+                nextMeleeTime = Time.time + meleeAttackData.duration;
+                animator.SetTrigger(Constants.MELEE_ATTACK_ANIMATION);
+            }
+        }
+    }
+
+    void FixedUpdate()
+    {
+        if (isAttacking || isFiringProjectile || player == null)
+        {
+            horizontalMove = 0f;
+        }
+        Move(horizontalMove * movement.maxSpeed * Time.fixedDeltaTime);
+    }
+
+    void Move(float move)
+    {
+        velocity = rb.velocity;
+        maxSpeedChange = movement.maxAcceleration * Time.fixedDeltaTime;
+		Vector3 desiredVelocity = new Vector2(move * 10f, rb.velocity.y);
+        velocity.x = Mathf.MoveTowards(velocity.x, desiredVelocity.x, maxSpeedChange);
+        rb.velocity = velocity;
+    }
+
+    private void Flip()
+	{
+		isFacingRight = !isFacingRight;
+		Vector3 localScale = transform.localScale;
+		localScale.x *= -1;
+		transform.localScale = localScale;
+	}
+
+    public void TakeDamage(AttackData attackData, Vector3 attackOrigin)
+    {
+        health.SubtractResource(attackData.damage);
+
+        if (health.GetCurrentValue() <= 0)
+        {
+            Die();
+        }
+    }
+
+    public void SpawnEnemyProjectile()
+    {
+        if (isFacingRight)
+        {
+            Instantiate(projectileAttackData.projectilePrefab, projectileSpawnPoint.position, transform.rotation);
+        }
+        else
+        {
+            Instantiate(projectileAttackData.projectilePrefab, projectileSpawnPoint.position, transform.rotation * Quaternion.Euler(0, 180, 0));
+        }
+    }
+
+    void Die()
+    {
+        GetComponent<BoxCollider2D>().enabled = false;
+        rb.velocity = Vector3.zero;
+        this.enabled = false;
+    }
+
+    IEnumerator FindingPlayer(float pollInterval) {
+        while (true)
+        {
+            player = FindPlayer();
+            yield return new WaitForSeconds(pollInterval);
+        }
+    }
+
+    PlayerMovementController FindPlayer()
+    {
+        if (player != null)
+        {
+            return player;
+        }
+        GameObject playerObj = GameObject.FindWithTag(Constants.PLAYER_TAG);
+        if (playerObj == null)
+        {
+            return null;
+        }
+        return playerObj.GetComponent<PlayerMovementController>();
+    }
```

Comment:
Since this is defined in multiple places, this might be a good candidate for abstracting out to a static helper class down the road. Fine for now, though.